src/sync/treiber_stack.rs

```diff
@@ -49,6 +49,20 @@ impl<T> TreiberStack<T> {
         self.try_pop()
     }
 
+    /// Swaps a stack with a new one.  Can be used to take all the
+    /// elements off the stack at once.
+    ///
+    /// Wait-free
+    pub fn swap(&self, new: TreiberStack<T>) -> TreiberStack<T> {
+        let guard = epoch::pin();
+        // new is exclusively owned and so can be written and read
+        // with relaxed atomics
+        let newhead = new.head.swap(None, Relaxed, &guard);
+        let head = self.head.swap_shared(newhead, AcqRel, &guard);
```

Comment:
Even though there is a data dependency, C/C++ memory model allows an execution where the instructions are executed as if they are reordered.  Note that data dependency plays a key role for only consume reads.  (Though consume's semantics is not yet clearly sorted out..)

I think the ordering should be indeed `Release`, in order to "publish" the data in the `new` stack in the `self.head` location.